Translate lifting-state-up.md (#238)
reactjs/ru.reactjs.org: content/docs/lifting-state-up.md

  - "docs/flux-todo-list.html"
---

> Часто несколько компонентов должны отражать изменения одних и тех же данных. Мы рекомендуем поднимать общее состояние до ближайшего общего предка. Давайте посмотрим, как это работает.

Reviewer: ```suggestion
Часто несколько компонентов должны отражать одни и те же изменяющиеся данные. Мы рекомендуем поднимать общее состояние до ближайшего общего предка. Давайте посмотрим, как это работает.
```

В этом разделе мы создадим калькулятор температуры, вычисляющий вскипит ли вода при заданной температуре.

> Мы начнем с компонента под названием `BoilingVerdict`. Он принимает температуру по шкале Цельсия в качестве пропа `celsius` и выводит, достаточна ли температура для кипения воды:

Reviewer 2: ```suggestion
Мы начнём с компонента под названием `BoilingVerdict`. Он принимает температуру по шкале Цельсия в качестве пропа `celsius` и выводит, достаточна ли температура для кипения воды:
```

}
```

[**Посмотреть на CodePen**](https://codepen.io/gaearon/pen/ZXeOBm?editors=0010)

## Добавление второго поля ввода {#adding-a-second-input}

> Наше новое требование состоит в том, что в дополнение к полю ввода градусов по шкале Цельсия мы добавляем аналогичное поле ввода, но по шкале Фаренгейта, и оба поля будут синхронизироваться.

Reviewer: ```suggestion
Наше новое требование состоит в том, что в дополнение к полю ввода температуры по шкале Цельсия мы добавляем аналогичное поле ввода, но по шкале Фаренгейта, и оба поля будут синхронизироваться.
```

Author: Выбрал вариант попроще из предложенных: https://github.com/reactjs/ru.reactjs.org/pull/238#discussion_r264375633

> Наше новое требование состоит в том, что в дополнение к полю ввода градусов по шкале Цельсия мы добавляем аналогичное поле ввода, но по шкале Фаренгейта, и оба поля будут синхронизироваться.

Reviewer 3: Попробовал покороче сделать, а то сложно читается.

```suggestion
Добавим к полю ввода градусов Цельсия, поле ввода по шкале Фаренгейта. Оба поля будут синхронизироваться.
```

}
```

[**Попробовать на CodePen**](https://codepen.io/gaearon/pen/jGBryx?editors=0010)

> Сейчас у нас есть два поля ввода, но когда вы вводите температуру в одно из них, другое поле не обновляется. Это противоречит нашему требованию: мы хотим их синхронизировать.

Reviewer 4: ```suggestion
Сейчас у нас есть два поля ввода, но когда вы вводите температуру в одно из них, другое поле не обновляется. Это противоречит нашему требованию — мы хотим их синхронизировать.
```

}
```

> Эти две функции преобразуют числа. Мы напишем ещё одну функцию, которая принимает строку с температурой (`temperature`) и функцию конвертации (`convert`) в качестве аргументов, и возвращает строку. Мы будем использовать эту функцию для вычисления значения одного поля ввода на основе значения из другого поля ввода.

Reviewer 3: Сверху уже был перевод – конвертация

```suggestion
Эти две функции конвертируют числа. Мы напишем ещё одну функцию, которая принимает строку с температурой (`temperature`) и функцию конвертации (`convert`) в качестве аргументов, и возвращает строку. Мы будем использовать эту функцию для вычисления значения одного поля ввода на основе значения из другого поля ввода.
```

}
```

> Например, вызов `tryConvert('abc', toCelsius)` возвратит пустую строку, а вызов `tryConvert('10.22', toFahrenheit)` возвращает `'50.396'`.

Reviewer 4: ```suggestion
Например, вызов `tryConvert('abc', toCelsius)` возвратит пустую строку, а вызов `tryConvert('10.22', toFahrenheit)` — `'50.396'`.
```

## Поднятие состояния {#lifting-state-up}

> В настоящее время оба компонента `TemperatureInput` независимо хранят свои значения в локальном состоянии:

Reviewer: ```suggestion
В настоящее время оба компонента `TemperatureInput` независимо хранят свои значения каждое в собственном локальном состоянии:
```

  render() {
    const temperature = this.state.temperature;
    // ...
```

> Однако мы хотим, чтобы эти два поля ввода синхронизировались друг с другом. Когда мы обновляем поле ввода градусов по Цельсию, поле ввода градусов по Фаренгейту должно отражать преобразованную температуру и наоборот.

Reviewer 3: ```suggestion
Однако мы хотим, чтобы эти два поля ввода синхронизировались друг с другом. Когда мы обновляем поле ввода градусов по Цельсию, поле ввода градусов по Фаренгейту должно отражать сконвертированную температуру и наоборот.
```

> В React совместное использование состояния достигается перемещением его до ближайшего предка компонентов, которым оно требуется. Это называется «подъём состояния». Мы удалим локальное состояние из `TemperatureInput` и переместим его в `Calculator`.

Reviewer 4: ```suggestion
В React совместное использование состояния достигается перемещением его до ближайшего предка компонентов, которым оно требуется. Это называется «подъём состояния». Мы удалим внутреннее состояние из `TemperatureInput` и переместим его в `Calculator`.
```

> Если `Calculator` владеет общим состоянием, он становится «источником истины» текущей температуры в обоих полей ввода. Он может предоставить им значения, которые не противоречат друг другу. Поскольку пропсы обоих компонентов `TemperatureInput` приходят из одного и того же родительского компонента `Calculator`, два поля ввода будут всегда синхронизированны.

Reviewer 2: ```suggestion
Если `Calculator` владеет общим состоянием, он становится «источником истины» текущей температуры для обоих полей ввода. Он может предоставить им значения, которые не противоречат друг другу. Поскольку пропсы обоих компонентов `TemperatureInput` приходят из одного и того же родительского компонента `Calculator`, два поля ввода будут всегда синхронизированны.
```

> Если `Calculator` владеет общим состоянием, он становится «источником истины» текущей температуры для обоих полей ввода. Он может предоставить им значения, которые не противоречат друг другу. Поскольку пропсы обоих компонентов `TemperatureInput` приходят из одного и того же родительского компонента `Calculator`, два поля ввода будут всегда синхронизированны.

Reviewer 4: ```suggestion
Если `Calculator` владеет общим состоянием, он становится «источником истины» текущей температуры для обоих полей ввода. Он может предоставить им значения, которые не противоречат друг другу. Поскольку пропсы обоих компонентов `TemperatureInput` приходят из одного и того же родительского компонента `Calculator`, два поля ввода будут всегда синхронизированы.
```

> Давайте посмотрим, как это работает шаг за шагом.

Reviewer 4: ```suggestion
Давайте шаг за шагом посмотрим, как это работает.
```

Во-первых, мы заменим `this.state.temperature` на `this.props.temperature` в компоненте `TemperatureInput`. Пока давайте представим, что `this.props.temperature` уже существует, хотя нам нужно будет передать его из `Calculator` в будущем:

```js{3}
  render() {
    // Ранее было так: const temperature = this.state.temperature;
    const temperature = this.props.temperature;
    // ...
```

> Мы знаем, что [пропсы доступны только для чтения](/docs/components-and-props.html#props-are-read-only). Когда `temperature` находилась в локальном состоянии, `TemperatureInput` мог просто вызвать `this.setState()` для изменения его значения. Однако теперь, когда `temperature` находится в родительском компоненте в качестве пропа, `TemperatureInput` не может контролировать его.

Reviewer 4: ```suggestion
Мы знаем, что [пропсы доступны только для чтения](/docs/components-and-props.html#props-are-read-only). Когда `temperature` находилась во внутреннем состоянии, `TemperatureInput` мог просто вызвать `this.setState()` для изменения его значения. Однако теперь, когда `temperature` находится в родительском компоненте в качестве пропа, `TemperatureInput` не может контролировать его.
```

> В React это обычно решается путём создания «управляемого» компонента. Точно так же, как DOM-элемент `<input>` принимает атрибуты `value` и `onChange`, то и пользовательский `TemperatureInput` принимает оба пропса `temperature` и `onTemperatureChange` от своего родителя `Calculator`.

Reviewer 4: ```suggestion
В React это обычно решается путём создания «управляемого» компонента. Точно так же, как DOM-элемент `<input>` принимает атрибуты `value` и `onChange`, так и пользовательский `TemperatureInput` принимает оба пропса `temperature` и `onTemperatureChange` от своего родителя `Calculator`.
```

> В React это обычно решается путём создания «контролируемого» компонента. Точно так же, как DOM-элемент `<input>` принимает атрибуты `value` и `onChange`, то и пользовательский `TemperatureInput` принимает оба пропса `temperature` и `onTemperatureChange` от своего родителя `Calculator`.

Reviewer 2: ```suggestion
В React это обычно решается путём создания «управляемого» компонента. Точно так же, как DOM-элемент `<input>` принимает атрибуты `value` и `onChange`, то и пользовательский `TemperatureInput` принимает оба пропса `temperature` и `onTemperatureChange` от своего родителя `Calculator`.
```

Author: Я оставил "контролируемый", т.к. этот термин из глоссария. Однако нашёл недавний коммит, где было изменено на "управляемый", так что поменяю на новый термин.

Теперь, когда `TemperatureInput` хочет обновить свою температуру, он вызывает `this.props.onTemperatureChange`:

```js{3}
  handleChange(e) {
    // Ранее было так: this.setState({temperature: e.target.value});
    this.props.onTemperatureChange(e.target.value);
    // ...
```

> Примечание:
>
> В пользовательских компонентах нет особого смысла в именах пропсов `temperature` или `onTemperatureChange`. Мы могли бы назвать их как-то иначе, например, `value` и` onChange`, т.к. подобные имена — распространённое соглашение.

> Пропсы `onTemperatureChange` и `temperature` будут предоставлены родительским компонентом `Calculator`. Он будет обрабатывать изменение модифицируя собственное локальное состояние, тем самым повторно отрендеривая оба поля ввода с новыми значениями. Мы вскоре рассмотрим новую реализацию `Calculator`.

Reviewer 4: ```suggestion
Пропсы `onTemperatureChange` и `temperature` будут предоставлены родительским компонентом `Calculator`. Он будет обрабатывать изменение, модифицируя собственное внутреннее состояние, тем самым повторно отрендеривая оба поля ввода с новыми значениями. Вскоре мы рассмотрим новую реализацию `Calculator`.
```

> Пропсы `onTemperatureChange` и `temperature` будут предоставлены родительским компонентом `Calculator`. Он будет обрабатывать изменение, модифицируя собственное внутреннее состояние, тем самым повторно отрендеривая оба поля ввода с новыми значениями. Вскоре мы рассмотрим новую реализацию `Calculator`.

Reviewer: ```suggestion
Пропсы `onTemperatureChange` и `temperature` будут предоставлены родительским компонентом `Calculator`. Он будет обрабатывать изменения, модифицируя собственное внутреннее состояние, тем самым повторно отрендеривая оба поля ввода с новыми значениями. Вскоре мы рассмотрим новую реализацию `Calculator`.
```

> Прежде чем погрузиться в изменения `Calculator`, давайте вспомним сделанные изменения в компоненте `TemperatureInput`. Мы удалили из него локальное состояние, и вместо использования `this.state.temperature` теперь используем `this.props.temperature`. Вместо вызова `this.setState()`, когда мы хотим внести изменения, теперь вызываем `this.props.onTemperatureChange()`, который передаётся компонентом `Calculator`:

Reviewer 4: ```suggestion
Прежде чем погрузиться в изменения `Calculator`, давайте вспомним сделанные изменения в компоненте `TemperatureInput`. Мы удалили из него внутреннее состояние, и вместо использования `this.state.temperature` теперь используем `this.props.temperature`. Вместо вызова `this.setState()`, когда мы хотим внести изменения, теперь вызываем `this.props.onTemperatureChange()`, который передаётся компонентом `Calculator`:
```

> Прежде чем погрузиться в изменения `Calculator`, давайте вспомним сделанные изменения в компонент `TemperatureInput`. Мы удалили из него локальное состояние, и вместо того использования `this.state.temperature` теперь используем `this.props.temperature`. Вместо вызова `this.setState()`, когда мы хотим внести изменения, теперь вызываем `this.props.onTemperatureChange()`, который передаётся компонентом `Calculator`:

Reviewer 2: ```suggestion
Прежде чем погрузиться в изменения `Calculator`, давайте вспомним сделанные изменения в компоненте `TemperatureInput`. Мы удалили из него локальное состояние, и вместо использования `this.state.temperature` теперь используем `this.props.temperature`. Вместо вызова `this.setState()`, когда мы хотим внести изменения, теперь вызываем `this.props.onTemperatureChange()`, который передаётся компонентом `Calculator`:
```

> Прежде чем погрузиться в изменения `Calculator`, давайте вспомним сделанные изменения в компоненте `TemperatureInput`. Мы удалили из него внутреннее состояние, и вместо использования `this.state.temperature` теперь используем `this.props.temperature`. Вместо вызова `this.setState()`, когда мы хотим внести изменения, теперь вызываем `this.props.onTemperatureChange()`, который передаётся компонентом `Calculator`:

Reviewer: ```suggestion
Прежде чем изменить `Calculator`, давайте вспомним, что поменялось в компоненте `TemperatureInput`. Мы удалили из него внутреннее состояние, и вместо `this.state.temperature` теперь используем `this.props.temperature`. Вместо вызова `this.setState()`, когда мы хотим изменить состояние, теперь вызываем `this.props.onTemperatureChange()`, который получен от компонента `Calculator`:
```

}
```

Теперь перейдём к компоненту `Calculator`.

> Мы будем хранить текущие значения `temperature` и `scale` в локальном состоянии. Это состояние, которое мы «подняли» от полей ввода, и теперь оно будет служить «источником истины» для них обоих. Это минимальное представление всех данных, про которое нам нужно знать для рендера обоих полей ввода.

Reviewer 4: ```suggestion
Мы будем хранить текущие значения `temperature` и `scale` во внутреннем состоянии. Это состояние, которое мы «подняли» от полей ввода, и теперь оно будет служить «источником истины» для них обоих. Это минимальное представление всех данных, про которое нам нужно знать для рендера обоих полей ввода.
```

Author: Можно поправить, но я бы добавил "во внутреннем состоянии этого компонента", т.к. непонятно в каком внутреннем состоянии мы это храним. И следующие предложения тоже не упоминают `Calculator`.

Reviewer 4: Обновил.

}
```

> [**Попробовать на CodePen**](https://codepen.io/gaearon/pen/WZpxpz?editors=0010)

Reviewer 4: ```suggestion
[**Посмотреть на CodePen**](https://codepen.io/gaearon/pen/WZpxpz?editors=0010)
```

Теперь, независимо от того, какое поле ввода вы редактируете, `this.state.temperature` и `this.state.scale` в `Calculator` обновляются. Одно из полей ввода получает значение как есть, поэтому введённые пользователем данные сохраняются, а значение другого поля ввода всегда пересчитывается на их основе.

Давайте посмотрим, что происходит, когда вы редактируете поле ввода:

* React вызывает функцию, указанную в `onChange` на DOM-элементе `<input>`. В нашем случае это метод `handleChange()` компонента `TemperatureInput`.
* Метод `handleChange()` в компоненте `TemperatureInput` вызывает `this.props.onTemperatureChange()` с новым требуемым значением. Его пропсы, включая `onTemperatureChange`, были предоставлены его родительским компонентом — `Calculator`.

> * Когда `Calculator` рендерился ранее, он указал, что `onTemperatureChange` в компоненте `TemperatureInput` по шкале Цельсия — это метод `handleCelsiusChange` в компоненте `Calculator`, а `onTemperatureChange` компонента `TemperatureInput` по шкале Фаренгейта — это метод `handleFahrenheitChange` в компоненте `Calculator`. Поэтому любой из этих двух методов `Calculator` вызывается в зависимости от того, какое поле ввода редактируется.

Reviewer: ```suggestion
* Когда `Calculator` рендерился ранее, он указал, что `onTemperatureChange` в компоненте `TemperatureInput` по шкале Цельсия — это метод `handleCelsiusChange` в компоненте `Calculator`, а `onTemperatureChange` компонента `TemperatureInput` по шкале Фаренгейта — это метод `handleFahrenheitChange` в компоненте `Calculator`. Поэтому один из этих двух методов `Calculator` вызывается в зависимости от того, какое поле ввода редактируется.
```

> * Внутри этих методов компонент `Calculator` указывает React сделать повторный рендер себя используя вызов `this.setState()` со значением нового поля ввода и текущей шкалой.

Reviewer 4: ```suggestion
* Внутри этих методов компонент `Calculator` указывает React сделать повторный рендер себя, используя вызов `this.setState()` со значением нового поля ввода и текущей шкалой.
```

* React вызывает метод `render()` компонента` Calculator`, чтобы узнать, как должен выглядеть UI. Значения обоих полей ввода пересчитываются исходя из текущей температуры и шкалы. В этом методе выполняется конвертация температуры.
* React вызывает методы `render()` конкретных компонентов `TemperatureInput` с их новыми пропсами, переданными компонентом `Calculator`. Он узнает, как должен выглядеть UI.
* React вызывает метод `render()` компонента `Boiling Verdict`, передавая температуру в градусах Цельсия как проп.

> * React DOM обновляет DOM, чтобы привести его в соответствие с нужными нам значениями в полях ввода. Отредактированное нами только что поле ввода получает его текущее значение, а другое поле ввода обновляется до температуры после конвертации.

Reviewer: ```suggestion
* React DOM обновляет DOM, чтобы привести его в соответствие с нужными нам значениями в полях ввода. Отредактированное нами только что поле ввода получает его текущее значение, а другое поле ввода обновляется конвертированным значением температуры.
```

Для подъёма состояния приходится писать больше «шаблонного» кода, чем при подходах с двусторонней привязкой данных, но мы получаем преимущество в виде меньших затрат на поиск и изолирование багов. Так как любое состояние «живёт» в каком-нибудь компоненте, и только этот компонент может его изменить, количество мест с возможными багами значительно уменьшается. Кроме того, вы можете реализовать любую пользовательскую логику для отклонения или преобразования данных, введённых пользователем.

Если что-то может быть вычислено из пропсов или из состояния, то скорее всего оно не должно находиться в состоянии. Например, вместо сохранения `celsiusValue` и `fahrenheitValue`, мы сохраняем только последнюю введённую температуру (`temperature`) и её шкалу (`scale`). Значение другого поля ввода можно всегда вычислить из них в методе `render()`. Это позволяет очистить или применить округление к значению другого поля, не теряя при этом точности значений, введённых пользователем.

> Когда вы видите, что в UI что-то не так, вы можете использовать [React Developer Tools](https://github.com/facebook/react-devtools) для проверки пропсов и навигации по дереву компонентов вверх до тех пор, пока не найдёте тот компонент, отвечающий за обновление состояния. Это позволяет отследить источник багов:

Reviewer 2: ```suggestion
Когда вы видите, что в UI что-то не так, вы можете использовать [React Developer Tools](https://github.com/facebook/react-devtools) для проверки пропсов и навигации по дереву компонентов вверх до тех пор, пока не найдёте тот компонент, который отвечает за обновление состояния. Это позволяет отследить источник багов:
```

> Когда вы видите, что в UI что-то не так, вы можете использовать [React Developer Tools](https://github.com/facebook/react-devtools) для проверки пропсов и навигации по дереву компонентов вверх до тех пор, пока не найдёте тот компонент, который отвечает за обновление состояния. Это позволяет отследить источник багов:

Reviewer 4: ```suggestion
Когда вы видите, что в интерфейсе происходит что-то не то, то можете воспользоваться расширением [React Developer Tools](https://github.com/facebook/react-devtools). С помощью него можно проверить пропсов и перемещаться по дереву компонентов вверх до тех пор, пока не найдёте тот компонент, который отвечает за обновление состояния. Это позволяет отследить источник багов:
```

Author: В TRANSLATION.md UI не только "аббревиатура, которую нельзя переводить", но и отдельная строка в глоссарии. В этом коммите есть несколько мест, к которым этот комментарий относится.

Reviewer 4: Верно, ранее мы переводили, ладно, потом может это изменим. А вам как лично кажется, нужно ли переводить?

Author: Оставлять как UI, если нет общего постоянного варианта сейчас. В предложенном изменении выше как-то несогласованно написано "..проверить пропсов..". 
В данном случае могу предложить переделать на "Если ваши компоненты отображают что-то не то... то вы можете воспользоваться... чтобы найти".

Reviewer 4: Обновил.

Author: "неладное" :-) Ладно, пока приму.

Reviewer 4: Другие варианты:
- странное
- непонятное
- невообразимое
- из ряда вон выходящее

Что-то нравится?

Author: Больше всего нравится первоначальный вариант "не так" или "в UI что-то отображается неправильно/некорректно, то..."

Reviewer 4: > "в UI что-то отображается неправильно/некорректно, то..."

Можете пожалуйста на это изменить? Просто хочется конкретики.

Author: Изменил
